### ISSUE TITLE
Potential race in *Work workflows when retrying/stopping work

### ISSUE THREAD

**Issue by MonsieurNicolas** [stellar/stellar-core#1706]
It looks like on retry (or destruction), the pattern followed is to typically call `clearChildren`.

Issue is that this does not actually stop children from running (references may still be held in callbacks), and may cause multiple instances of the same "Work" to run (potentially conflicting with each other).

There are a couple of challenges that need to be overcome here:
* there should be new sematics to stop or abort `Work` - such transition may be asynchronous in nature and may require additional state transitions in the various `Work` implementations (so that parents properly wait for all of their children to be done aborting work).
   * most `Work` implementations should probably handle "success" from a child work differently if they are themselves in aborted state (to avoid run away work to trigger more work for example).
* synchronous stopping such as in destructors, should "just work" and be safe
   * this is especially challenging as `RunCommandWork` uses `ProcessManager` and process termination is likely to stay asynchronous (ie, relies on an event posted on the main IOService).
   * It may be possible to "force kill" processes in this case to at least ensure that resources are freed relatively quickly
   * `RunCommandWork` (and others) may have to behave a little bit differently in that notification to parents should not occur (as parents may be destructed or partially destructed) 
   * as "zombie" processes may still run for a bit we may have to add additional logic before spawning new instances of related work to avoid conflicts (using unique folders per instance may be enough - but might be too error prone). For example: a new "catchup" work may trigger a bunch of child work that conflicts with old child work from a previous (aborted) run.

**Comment by marta-lokhova** [comment]
Here are some thoughts/ideas on asynchronous `abort` for Work classes:

High level: when parent is notified of child’s failure, a few things must happen:
* Parent goes into "aborting" state and “Abort” signal is issued to all other children, which in turn issue the signal to their children and so on until everybody knows that they are being aborted.
* Parent does not proceed (i.e. does not scheduleAbort/Failure) until all children move to aborted (or completed, see explanation below) state 
* Once the above happens, parent is scheduled to either abort itself or fail, propagating failure state from the child that triggered abort 

### Implementation Details
* Two new states
    * WORK_ABORTING 
    * WORK_FAILURE_ABORTED [work finished aborting]
* `abort` and `scheduleAbort` methods
    * `scheduleAbort`
        * Similarly to `scheduleRun` or `scheduleComplete`, schedules the work to complete aborting 
        * Essentially, posts callback for `complete` with either WORK_FAILURE_ABORTED or WORK_FAILURE_RAISE/WORK_FAILURE_FATAL failure states 
    * `abort` ensures signal is propagated to the children before scheduling parent itself to abort 
        * Sets state to WORK_ABORTING 
        * Checks if there are any unaborted children. If so, call `abort` on each child 
            * By “unaborted” we mean WORK_PENDING. Why pending?
                * If child work has finished already, it could be in 3 states: success, failure (nothing to do for either) or aborted
                * [*]  If work is in WORK_RUNNING state, don’t do anything and let it finish, it will notify parent regardless and parent can handle it in `advance`
       * Checks if any children are not done. If so, return (see [*])
       * Else calls `scheduleAbort`
            * Which means all children are completed or aborted, so the work itself if ready to abort 
* `complete` also needs to be updated to handle WORK_FAILURE_ABORTED state and properly notify parent 
* Need to ensure that new children are not added if parent it in WORK_ABORTING  state
    * e.g. `BatchWork` shouldn't yield any new work
* `advance` method also needs to be updated, as it’s called by notify
   * Pseudocode (tentative, might get simplified also)
```
if (state == WORK_ABORTING  && allPendingChildrenAborted())
{
	if allChildrenAborted() or allChildrenSuccess()
		scheduleFailure(state=WORK_FAILURE_ABORTED)
	else if anyFatalFailures()
		scheduleFailure(state=WORK_FAILURE_FATAL)
	else if anyRaiseFailures()
		scheduleFailure(state=WORK_FAILURE_RAISE)
}
if (state != WORK_ABORTING )
{
	if allChildrenSuccess()
		scheduleSuccess()
	if anyFatalFailures() or anyRaiseFailures()
		abort()
}
```
### Things to keep in mind
* Since we are waiting for RUNNING children to complete, do we need a timeout on some commands? What’s the max amount of time we need to wait?

**Comment by MonsieurNicolas** [comment]
cool.

A few questions:
1. Can you detail the split in terms of responsibilities for the caller? Specifically the interface changes at `WorkParent` vs `Work` level?
For example, I imagine that `abort` gets exposed at `WorkParent` level, anything else?

2. Are there big changes that need to be done to the process manager? (used by `RunCommand`)

3. What about the second bullet point?
> synchronous stopping such as in destructors, should "just work" and be safe

4. Yes, I imagine that there is some mechanism for timing out. In which case it might be the same behavior than what we get for dealing with the "force kill" required for the destructor case (bullet 3)

**Comment by marta-lokhova** [comment]
1. Yes. Additionally, `scheduleAbort` would be in `Work`. `complete` and `advance` would stay in `Work` also. 

2. Correct me if I am wrong, but if we wait for any running work to finish before aborting, we shouldn't need to modify process manager at that stage (first bullet). I see that `RunCommand` calls process manager in `onStart`, which is called when Work is actually being run. So if it's already running, we let it finish, otherwise it will be scheduled to fail without going into `onStart`.

3. Like I mentioned, the comment above is just to cover the first bullet. I'm working out the logic for the second bullet. There are a lot of details to consider there, but I'll post here once those are done.

**Comment by marta-lokhova** [comment]
@MonsieurNicolas 

> * this is especially challenging as RunCommandWork uses ProcessManager and process termination is likely to stay asynchronous (ie, relies on an event posted on the main IOService).
> * It may be possible to "force kill" processes in this case to at least ensure that resources are freed relatively quickly

* Actually it looks like there's shutdown logic implemented in `ProcessManagerImpl` with `shutdown`, `cleanShutdown` and `forceShutdown` methods. It appears that when Application is being destroyed, it properly shuts down all pending and running processes (see `ApplicationImpl::~ApplicationImpl`). 

* I guess some catchup-related processes can be shutdown in the destructor of `RunCommand` specifically (instead of letting Application handle it on shutdown), but I'm not sure when this would be needed?

> RunCommandWork (and others) may have to behave a little bit differently in that notification to parents should not occur (as parents may be destructed or partially destructed)

I'm trying to think when this would happen. If running work is destructed in proper order, this shouldn't cause issues, should it? 

> as "zombie" processes may still run for a bit we may have to add additional logic before spawning new instances of related work to avoid conflicts (using unique folders per instance may be enough - but might be too error prone). For example: a new "catchup" work may trigger a bunch of child work that conflicts with old child work from a previous (aborted) run.

So if application is re-started, we should not have any "zombie" processes. If application is running, but catchup work failed and is retrying, with the new abort semantics, we should not find ourselves chasing ghost processes, as previous catchup would either wait for `RunCommand` to finish or fail it before it even spawns a new process. 

Let me know what you think.

**Comment by marta-lokhova** [comment]
It does seem cleaner if we have the guarantee that work can only be destroyed if it's in a completed state. Some more thoughts on what we could do:

1. `ProcessManager` waits for abort of `WorkManager`
* In`ApplicationImpl::gracefulStop` and in `ApplicationImpl::~ApplicationImpl()`, an additional command to abort `WorkManager` is issued prior to `ProcessManager` shutdown. This top level abort would just perform graceful abort that I described in the first comment. 
* That should correspond ok with the ProcessManager, as we would wait for anything that's in the process queue or running, and abort anything that hasn't been put in the queue yet. 
* Meanwhile, other parts of the system might still be running and trying to add work. We should introduce some rules on the `Work` level; specifically, that any attempts to add new work are ignored if Work is being aborted. 
* `WorkManager` abort is asynchronous, and we should ensure not to shutdown ProcessManager until WorkManager is done aborting. Perhaps it'd need to asynchronously wait for WM.

2. Proper timeout in `RunCommand` to avoid long waits
* For that, I think `ProcessExitEvent` needs to be updated (currently it has a timer that doesn't expire, and includes both time in the queue + execution time). To narrow it down to just execution time, we might need another timer, that sets once we reach `ProcessExitEvent::Impl::run()` (new process is started) and properly cancels timed out process. Such process should also be force killed. 
* I'm not entirely sure what's a reasonable number to wait (due to varying sizes of buckets + potential network slowdown), you might have a better idea
* `RunCommand` would then handle timer expiry as a regular failure, and proceed as usual (notifying parent of failure etc). We might want to add another state to differentiate, say, `WORK_FAILURE_TIMEOUT` (we'd deduce this state from the error code `asio::error::operation_aborted`)

**Comment by marta-lokhova** [comment]
In #1729, @vogel brought up a good point that we might want to abort work that is not just in PENDING state, but in RUNNING state as well. The advantage is that it saves a lot of time; if we are aborting during bucket application, we don't need to wait for [potentially very long] work to finish. The disadvantage is that it adds complexity (and some tricky edge cases), since now Work needs to interact with the ProcessManager to figure out which processes are running or in a process queue and how to handle them. Prime example is `RunCommandWork`: it relies on `ProcessExitEvent` to trigger work completion. 

My proposition on what we can do to abort a running work:
* Add a new method `ProcessManager::shutdownProcess(ProcessExitEvent pev)`, which similarly to shutdown logic, removes process from pending queue if it's not running yet, and sends kill signal and removes process from running queue otherwise.
* Override `onAbort` for `RunProcessCommand` so that it could call the method above.
* Update `Work::callComplete` to handle abort case

**Comment by marta-lokhova** [comment]
Resolved by #2063
